src/libstd/ffi/c_str.rs

```diff
@@ -255,6 +255,34 @@ impl CString {
         Box::into_raw(self.inner) as *mut libc::c_char
     }
 
+    /// Converts the `CString` into a `String` if it contains valid Unicode data.
+    ///
+    /// On failure, ownership of the original `CString` is returned.
+    #[unstable(feature = "convert", reason = "recently added", issue = "27704")]
+    pub fn into_string(self) -> Result<String, CString> {
+        String::from_utf8(self.into_bytes())
+            .map_err(|e| unsafe { CString::from_vec_unchecked(e.into_bytes()) })
+    }
+
+    /// Returns the underlying byte buffer.
+    ///
+    /// The returned buffer does **not** contain the trailing nul separator and
+    /// it is guaranteed to not have any interior nul bytes.
+    #[unstable(feature = "convert", reason = "recently added", issue = "27704")]
+    pub fn into_bytes(self) -> Vec<u8> {
```

Comment:
Mm, my only concern was that it's not necessarily straightforward whether you expect the null byte to be included or not -- and we can't really document trait impls (or at least, not in a way that they'd show up in rustdoc). For the record, I find the behaviour of trimming off the null byte surprising for `CStr::as_bytes`, and mainly only realised it did because I happened to notice the `as_bytes_with_nul` method alongside it.

Also, it feels inconsistent compared to other similar structures. Would it be worth deprecating, for example, `String::into_bytes` and other similar methods in that case and try to push `Into` as the proper way to do these conversions that can't fail?